File: 0117 - Populating Next Right Pointers in Each Node II/solution.c
/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     struct Node *left;
 *     struct Node *right;
 *     struct Node *next;
 * };
 */
struct Node* findNext(struct Node *root){
    if(!root) return NULL;
    if(root->left) return root->left;
    else if(root->right) return root->right;
    else return findNext(root->next);
}
struct Node* connect(struct Node *root){
	if(!root) return NULL;
    if(root->left) root->left->next = root->right?root->right:findNext(root->next);
    if(root->right) root->right->next = findNext(root->next);
    connect(root->right);
    connect(root->left);
    return root;
}

File: 0117 - Populating Next Right Pointers in Each Node II/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        head, p, self.next = root, self, None
        while head:
            if head.left:
                p.next = head.left
                p = p.next
            if head.right:
                p.next = head.right
                p = p.next
            if head.next:
                head = head.next
            else:
                head = self.next
                p, self.next = self, None
        return root